addon/helpers/date-diff.js

```diff
@@ -0,0 +1,52 @@
+import { helper } from '@ember/component/helper';
+import {
+  differenceInDays,
+  differenceInHours,
+  differenceInMinutes,
+  differenceInSeconds,
+} from 'date-fns';
+import normalizeDate from '../utils/normalize-date';
+import { assert } from '@ember/debug';
+
+/**
+  Return the difference between two given dates as a number.
+  If you want a formatted response, see: date-format or date-from-now
+
+  @param {Date|String|Number} dateA given date
+  @param {Date|String|Number} dateB given date
+  @param {String} inputFormat string of tokens representing a date format
+  @param {String} precision a unit of precision for the return value
+  @return {Number|Null} a positive integer representing the difference, or null if invalid
+*/
+export default helper(function dateDiff([dateA, dateB], { precision = 'days', inputFormat }) {
+  if (!dateA || !dateB) return null;
+
+  assert(
+    '{{date-diff helper should not accept date args of mixed type',
+    typeof dateA === typeof dateB
+  );
```

Comment:
Regarding this change, cc @jherdman @christophermlne 

It occurred to me that perhaps Typescript would save us from having to do something like this.. but i'm not actually sure if TS can evaluate different args and assert like that?

Reply:
I'm a little out of my depth here too. I'm hoping to begin porting some addons to TS soon to get a better understanding of how it can help us with this sort of problem.